File: src/test/java/com/codeperfection/shipit/controller/ProductControllerTest.java
package com.codeperfection.shipit.controller;

import com.codeperfection.shipit.dto.common.PaginationFilterDto;
import com.codeperfection.shipit.dto.product.CreateProductDto;
import com.codeperfection.shipit.dto.product.ProductDto;
import com.codeperfection.shipit.dto.product.UpdateCountInStockDto;
import com.codeperfection.shipit.dto.product.UpdateProductDto;
import com.codeperfection.shipit.exception.errordto.ErrorType;
import com.codeperfection.shipit.service.ProductService;
import com.codeperfection.shipit.util.AuthenticationFixtureFactory;
import com.codeperfection.shipit.util.CommonFixtureFactory;
import com.codeperfection.shipit.util.ProductFixtureFactory;
import com.codeperfection.shipit.util.TestUtil;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;

import java.util.UUID;

import static com.codeperfection.shipit.controller.CommonPathValues.API_V1;
import static com.codeperfection.shipit.controller.ProductController.COUNT_IN_STOCK_PATH;
import static com.codeperfection.shipit.controller.ProductController.PRODUCTS_PATH;
import static org.hamcrest.Matchers.containsInAnyOrder;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

public class ProductControllerTest extends ControllerTestBase {

    @MockBean
    private ProductService productService;

    @Test
    public void createProduct_IfNotAuthenticated_ReturnsError() throws Exception {
        checkUnauthorizedResponse(post(API_V1 + PRODUCTS_PATH));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void createProduct_IfInvalidPayload_ReturnsError() throws Exception {
        mockAuthentication();
        final var createProductDto = new CreateProductDto("", 0, 0, -1);
        mockMvc.perform(post(API_V1 + PRODUCTS_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createProductDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is4xxClientError())
                .andExpect(jsonPath("$.errorType", is(ErrorType.INVALID_REQUEST.getDisplayName())))
                .andExpect(jsonPath("$.fieldErrors[*].fieldName",
                        containsInAnyOrder("name", "volume", "price", "countInStock")));
        verifyNoInteractions(productService);
    }

    @Test
    public void createProduct_IfValidPayload_ReturnsDto() throws Exception {
        mockAuthentication();
        final var createProductDto = ProductFixtureFactory.createCreateProductDto();
        final var productDto = ProductFixtureFactory.createProductDto();
        final var authenticatedUser = AuthenticationFixtureFactory.createAuthenticatedUser();
        doReturn(productDto).when(productService).createProduct(createProductDto, authenticatedUser);

        mockMvc.perform(post(API_V1 + PRODUCTS_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(productDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().json(objectMapper.writeValueAsString(productDto)))
                .andExpect(redirectedUrlPattern("http://*" + API_V1 + PRODUCTS_PATH + "/" + productDto.getUuid()));

        verify(productService).createProduct(createProductDto, authenticatedUser);
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void getProducts_IfNotAuthenticated_ReturnsError() throws Exception {
        checkUnauthorizedResponse(get(API_V1 + PRODUCTS_PATH));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void getProducts_IfInvalidPagination_ReturnsError() throws Exception {
        mockAuthentication();
        final var invalidPaginationFilterDto = new PaginationFilterDto(-1, 0);
        mockMvc.perform(get(API_V1 + PRODUCTS_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .params(TestUtil.toMultiValueMap(invalidPaginationFilterDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is4xxClientError())
                .andExpect(jsonPath("$.errorType", is(ErrorType.INVALID_REQUEST.getDisplayName())))
                .andExpect(jsonPath("$.fieldErrors[*].fieldName",
                        containsInAnyOrder("page", "size")));
        verifyNoInteractions(productService);
    }

    @Test
    public void getProducts_IfValidPagination_ReturnsPageDto() throws Exception {
        mockAuthentication();
        final var paginationFilterDto = CommonFixtureFactory.createPaginationFilterDto();
        final var productDto = ProductFixtureFactory.createProductDto();
        final var pageDto = CommonFixtureFactory.createPageDto(productDto);
        doReturn(pageDto).when(productService).getProducts(paginationFilterDto, authenticatedUser);
        mockMvc.perform(get(API_V1 + PRODUCTS_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .params(TestUtil.toMultiValueMap(paginationFilterDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().json(objectMapper.writeValueAsString(pageDto)));
        verify(productService).getProducts(paginationFilterDto, authenticatedUser);
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void getProduct_IfNotAuthenticated_ReturnsError() throws Exception {
        final var productUuid = UUID.fromString("86bc3ac7-7ba5-446c-a751-9a525f7b2378");
        checkUnauthorizedResponse(get(API_V1 + PRODUCTS_PATH + "/" + productUuid));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void getProduct_IfInvalidPathVariable_ReturnsError() throws Exception {
        checkBadRequestResponseOnInvalidPathVariable(get(API_V1 + PRODUCTS_PATH + "/" +
                "InvalidUuid86bc3ac7-7ba5-446c-a751-9a525f7b2378"));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void getProduct_IfValidUuid_ReturnsDto() throws Exception {
        mockAuthentication();
        final var productDto = ProductFixtureFactory.createProductDto();
        doReturn(productDto).when(productService).getProduct(productDto.getUuid(), authenticatedUser);
        mockMvc.perform(get(API_V1 + PRODUCTS_PATH + "/" + productDto.getUuid())
                .contentType(MediaType.APPLICATION_JSON)
                .param("uuid", productDto.getUuid().toString())
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().json(objectMapper.writeValueAsString(productDto)));
        verify(productService).getProduct(productDto.getUuid(), authenticatedUser);
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateProduct_IfNotAuthenticated_ReturnsError() throws Exception {
        final var productUuid = UUID.fromString("86bc3ac7-7ba5-446c-a751-9a525f7b2378");
        checkUnauthorizedResponse(put(API_V1 + PRODUCTS_PATH + "/" + productUuid));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateProduct_IfInvalidPathVariable_ReturnsError() throws Exception {
        checkBadRequestResponseOnInvalidPathVariable(put(API_V1 + PRODUCTS_PATH + "/" +
                "InvalidUuid86bc3ac7-7ba5-446c-a751-9a525f7b2378"));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateProduct_IfInvalidPayload_ReturnsError() throws Exception {
        mockAuthentication();
        final var productUuid = UUID.randomUUID();
        final var updateProductDto = new UpdateProductDto("", -1, -1);

        mockMvc.perform(put(API_V1 + PRODUCTS_PATH + "/" + productUuid.toString())
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(updateProductDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is4xxClientError())
                .andExpect(jsonPath("$.errorType", is(ErrorType.INVALID_REQUEST.getDisplayName())))
                .andExpect(jsonPath("$.fieldErrors[*].fieldName",
                        containsInAnyOrder("name", "volume", "price")));

        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateProduct_IfValidPayload_ReturnsDto() throws Exception {
        mockAuthentication();
        final var productUuid = UUID.randomUUID();
        final var updateProductDto = new UpdateProductDto("a", 1, 1);
        final var productDto = new ProductDto(UUID.randomUUID(), updateProductDto.getName(),
                updateProductDto.getVolume(), updateProductDto.getPrice(), 5);
        doReturn(productDto).when(productService).updateProduct(productUuid, updateProductDto, authenticatedUser);

        mockMvc.perform(put(API_V1 + PRODUCTS_PATH + "/" + productUuid.toString())
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(updateProductDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().json(objectMapper.writeValueAsString(productDto)))
                .andExpect(redirectedUrlPattern("http://*" + API_V1 + PRODUCTS_PATH + "/" + productDto.getUuid()));

        verify(productService).updateProduct(productUuid, updateProductDto, authenticatedUser);
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateCountInStock_IfNotAuthenticated_ReturnsError() throws Exception {
        final var productUuid = UUID.fromString("86bc3ac7-7ba5-446c-a751-9a525f7b2378");
        checkUnauthorizedResponse(put(API_V1 + PRODUCTS_PATH + "/" + productUuid + COUNT_IN_STOCK_PATH));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateCountInStock_IfInvalidPathVariable_ReturnsError() throws Exception {
        checkBadRequestResponseOnInvalidPathVariable(put(API_V1 + PRODUCTS_PATH + "/" +
                "InvalidUuid86bc3ac7-7ba5-446c-a751-9a525f7b2378" + COUNT_IN_STOCK_PATH));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateCountInStock_IfInvalidPayload_ReturnsError() throws Exception {
        mockAuthentication();
        final var productUuid = UUID.randomUUID();
        final var updateCountInStockDto = new UpdateCountInStockDto(-1);

        mockMvc.perform(put(API_V1 + PRODUCTS_PATH + "/" + productUuid.toString() + COUNT_IN_STOCK_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(updateCountInStockDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is4xxClientError())
                .andExpect(jsonPath("$.errorType", is(ErrorType.INVALID_REQUEST.getDisplayName())))
                .andExpect(jsonPath("$.fieldErrors[*].fieldName",
                        containsInAnyOrder("countInStock")));

        verifyNoMoreInteractions(productService);
    }

    @Test
    public void updateCountInStock_IfValidPayload_ReturnsDto() throws Exception {
        mockAuthentication();
        final var productUuid = UUID.randomUUID();
        final var updateCountInStockDto = new UpdateCountInStockDto(3);
        final var productDto = new ProductDto(productUuid, "name", 1, 1,
                updateCountInStockDto.getCountInStock());
        doReturn(productDto).when(productService).updateCountInStock(
                productUuid, updateCountInStockDto, authenticatedUser);

        mockMvc.perform(put(API_V1 + PRODUCTS_PATH + "/" + productUuid.toString() + COUNT_IN_STOCK_PATH)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(updateCountInStockDto))
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful())
                .andExpect(content().json(objectMapper.writeValueAsString(productDto)));

        verify(productService).updateCountInStock(productUuid, updateCountInStockDto, authenticatedUser);
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void deleteProduct_IfNotAuthenticated_ReturnsError() throws Exception {
        final var productUuid = UUID.fromString("86bc3ac7-7ba5-446c-a751-9a525f7b2378");
        checkUnauthorizedResponse(delete(API_V1 + PRODUCTS_PATH + "/" + productUuid + COUNT_IN_STOCK_PATH));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void deleteProduct_IfInvalidPathVariable_ReturnsError() throws Exception {
        checkBadRequestResponseOnInvalidPathVariable(delete(API_V1 + PRODUCTS_PATH + "/" +
                "InvalidUuid86bc3ac7-7ba5-446c-a751-9a525f7b2378"));
        verifyNoMoreInteractions(productService);
    }

    @Test
    public void deleteProduct_ReturnsSuccessfulStatus() throws Exception {
        mockAuthentication();
        final var productUuid = UUID.randomUUID();
        mockMvc.perform(delete(API_V1 + PRODUCTS_PATH + "/" + productUuid.toString())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, getJwtMockAuthorization()))
                .andExpect(status().is2xxSuccessful());

        verify(productService).deleteProduct(productUuid, authenticatedUser);
        verifyNoMoreInteractions(productService);
    }
}

File: src/test/kotlin/com/codeperfection/shipit/controller/ProductControllerTest.kt
package com.codeperfection.shipit.controller

import com.codeperfection.shipit.dto.common.PageDto
import com.codeperfection.shipit.dto.common.PaginationFilterDto
import com.codeperfection.shipit.dto.product.UpdateProductDto
import com.codeperfection.shipit.exception.clienterror.NotFoundException
import com.codeperfection.shipit.exception.dto.ErrorType
import com.codeperfection.shipit.fixture.*
import com.codeperfection.shipit.service.ProductService
import org.hamcrest.Matchers
import org.hamcrest.Matchers.hasSize
import org.json.JSONArray
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

class ProductControllerTest : ControllerTestBase() {

    @MockBean
    private lateinit var productService: ProductService

    @AfterEach
    fun tearDown() {
        Mockito.verifyNoMoreInteractions(productService)
    }

    private val productJson = """
        {
            "id": "$PRODUCT_ID_1",
            "userId": "$USER_ID",
            "name": "$PRODUCT_NAME_1",
            "volume": $PRODUCT_VOLUME_1,
            "price": $PRODUCT_PRICE_1,
            "countInStock": $PRODUCT_COUNT_IN_STOCK_1
        }
    """.trimIndent()

    @Test
    @WithMockUser
    fun `GIVEN invalid request, WHEN creating product, THEN error response is returned`() {
        mockMvc.perform(
            post("/api/v1/users/$USER_ID/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "name": "  ",
                        "volume": 0,
                        "price": 0,
                        "countInStock": -1
                    }
                    """.trimIndent()
                )
        )
            .andExpect(status().isBadRequest)
            .andExpect(jsonPath("errorType", Matchers.`is`(ErrorType.INVALID_REQUEST.name)))
            .andExpect(jsonPath("fieldErrors", hasSize<JSONArray>(4)))
    }

    @Test
    @WithMockUser
    fun `GIVEN valid request, WHEN creating product, THEN created product is returned`() {
        whenever(productService.createProduct(USER_ID, createProductDtoFixture1)).thenReturn(productDtoFixture1)

        mockMvc.perform(
            post("/api/v1/users/$USER_ID/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "name": "$PRODUCT_NAME_1",
                        "volume": $PRODUCT_VOLUME_1,
                        "price": $PRODUCT_PRICE_1,
                        "countInStock": $PRODUCT_COUNT_IN_STOCK_1
                    }
                    """.trimIndent()
                )
        )
            .andExpect(status().isCreated)
            .andExpect(content().json(productJson))
            .andExpect(
                header().string(
                    "location", "http://localhost/api/v1/users/$USER_ID/products/$PRODUCT_ID_1"
                )
            )

        verify(productService).createProduct(USER_ID, createProductDtoFixture1)
    }

    @Test
    @WithMockUser
    fun `GIVEN invalid request, WHEN getting products page, THEN error response is returned`() {
        mockMvc.perform(
            get("/api/v1/users/$USER_ID/products")
                .queryParam("page", "-1")
                .queryParam("size", "0")
        )
            .andExpect(status().isBadRequest)
            .andExpect(jsonPath("errorType", Matchers.`is`(ErrorType.INVALID_REQUEST.name)))
            .andExpect(jsonPath("fieldErrors", hasSize<JSONArray>(2)))
    }

    @Test
    @WithMockUser
    fun `GIVEN valid request, WHEN getting products page, THEN products page is returned`() {
        val paginationFilterDto = PaginationFilterDto(page = 0, size = 1)
        whenever(productService.getProducts(USER_ID, paginationFilterDto))
            .thenReturn(PageDto(totalElements = 1, totalPages = 1, elements = listOf(productDtoFixture1)))

        mockMvc.perform(
            get("/api/v1/users/$USER_ID/products")
                .queryParam("page", "0")
                .queryParam("size", "1")
        )
            .andExpect(status().isOk)
            .andExpect(
                content().json(
                    """
                    {
                        "totalElements": 1,
                        "totalPages": 1,
                        "elements": [
                            $productJson
                        ]
                    }
                    """.trimIndent()
                )
            )

        verify(productService).getProducts(USER_ID, paginationFilterDto)
    }

    @Test
    @WithMockUser
    fun `GIVEN request for a product that doesn't exist, WHEN getting the product, THEN not found is returned`() {
        whenever(productService.getProduct(USER_ID, PRODUCT_ID_1)).thenThrow(NotFoundException(PRODUCT_ID_1, USER_ID))

        mockMvc.perform(get("/api/v1/users/$USER_ID/products/$PRODUCT_ID_1"))
            .andExpect(status().isNotFound)
            .andExpect(jsonPath("errorType", Matchers.`is`(ErrorType.NOT_FOUND.name)))

        verify(productService).getProduct(USER_ID, PRODUCT_ID_1)
    }

    @Test
    @WithMockUser
    fun `GIVEN request for a product that exists, WHEN getting the product, THEN product is returned`() {
        whenever(productService.getProduct(USER_ID, PRODUCT_ID_1)).thenReturn(productDtoFixture1)

        mockMvc.perform(get("/api/v1/users/$USER_ID/products/$PRODUCT_ID_1"))
            .andExpect(status().isOk)
            .andExpect(content().json(productJson))

        verify(productService).getProduct(USER_ID, PRODUCT_ID_1)
    }

    @Test
    @WithMockUser
    fun `GIVEN invalid request, WHEN updating product, THEN error response is returned`() {
        mockMvc.perform(
            put("/api/v1/users/$USER_ID/products/$PRODUCT_ID_1")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "countInStock": -1
                    }
                    """.trimIndent()
                )
        )
            .andExpect(status().isBadRequest)
            .andExpect(jsonPath("errorType", Matchers.`is`(ErrorType.INVALID_REQUEST.name)))
            .andExpect(jsonPath("fieldErrors", hasSize<JSONArray>(1)))
    }

    @Test
    @WithMockUser
    fun `GIVEN valid request, WHEN updating product, THEN updated product is returned`() {
        val newCountInStock = 123
        whenever(productService.updateProduct(USER_ID, PRODUCT_ID_1, UpdateProductDto(countInStock = newCountInStock)))
            .thenReturn(productDtoFixture1.copy(countInStock = newCountInStock))

        mockMvc.perform(
            put("/api/v1/users/$USER_ID/products/$PRODUCT_ID_1")
                .contentType(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "countInStock": $newCountInStock
                    }
                    """.trimIndent()
                )
        )
            .andExpect(status().isOk)
            .andExpect(
                content().json(
                    """
                    {
                        "id": "$PRODUCT_ID_1",
                        "userId": "$USER_ID",
                        "name": "$PRODUCT_NAME_1",
                        "volume": $PRODUCT_VOLUME_1,
                        "price": $PRODUCT_PRICE_1,
                        "countInStock": $newCountInStock
                    }
                    """.trimIndent()
                )
            )

        verify(productService).updateProduct(USER_ID, PRODUCT_ID_1, UpdateProductDto(countInStock = newCountInStock))
    }

    @Test
    @WithMockUser
    fun `GIVEN valid request, WHEN deleting product, THEN no content is returned`() {
        mockMvc.perform(delete("/api/v1/users/$USER_ID/products/$PRODUCT_ID_1"))
            .andExpect(status().isNoContent)

        verify(productService).deleteProduct(USER_ID, PRODUCT_ID_1)
    }
}
